### DISCUSSION TITLE
Nextjs Prod build error

### DISCUSSION THREAD

**Discussion by Sathyaraj-dev** [vercel/next.js#66268] [App Router] [answered]
TypeError: fetch failed
    at node:internal/deps/undici/undici:12345:11
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
  cause: AggregateError
      at internalConnectMultiple (node:net:1114:18)
      at afterConnectMultiple (node:net:1667:5)
      at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
    code: 'ECONNREFUSED',
    [errors]: [ [Error], [Error] ]
  },
  digest: '1990109165'
}

Anyone know why this error is coming on npm run build? Please help me on this.

**Comment by icyJoseph** [comment, marked as answer]
Hi,

You are most likely fetching your own server's API endpoint, but during build time there's no API endpoint, because the server is not build just yet.

- https://nextjs-faq.com/fetch-api-in-rsc

Instead of fetching from your own API endpoint, inline the code that gets the data you need, into your server component, which is one of the reasons to have server components in the first place.

The link above summarizes the problem, please do read through it.

**Reply by Sathyaraj-dev** [reply]
Thanks, icyJoseph. Yes, I called the API on server components, and that's the issue. I cross-checked the production checklist and found the answers.

[Route Handlers](https://nextjs.org/docs/app/building-your-application/routing/route-handlers): Use Route Handlers to access your backend resources from Client Components. But do not call Route Handlers from Server Components to avoid an additional server request.

Thanks for your help!

**Reply by PunkFleet** [reply]
> Hi,
> 
> You are most likely fetching your own server's API endpoint, but during build time there's no API endpoint, because the server is not build just yet.
> 
> * https://nextjs-faq.com/fetch-api-in-rsc
> 
> Instead of fetching from your own API endpoint, inline the code that gets the data you need, into your server component, which is one of the reasons to have server components in the first place.
> 
> The link above summarizes the problem, please do read through it.

Hi, Im newbie for Nextjs, I moved from vitejs to nextjs and am trying to build rewritten projects. I read the article but case in point it seems to use a prisma plugin. Can I not use it?

**Comment by PunkFleet** [comment]
@icyJoseph 
I have a server `src/app/name/page.tsx` and currently it's requesting my backend api directly within the component
```

type Props = {
    params: { name: string; lang: string };
};

export async function generateMetadata({ params }: Props, parent: ResolvingMetadata): Promise<Metadata> {
    const item = await getPublicData("name", params.name);
    return {
        title: `${params.name}.`,
        keywords: item.seo_title,
        description: item.seo_description,
    };
}
export default async function Page({ params }: { params: { name: string; lang: string } }) {
        const item = await getPublicData('name', params.name)
    const features = await getPublicData("features")
    const faq = await getPublicData("faq_detail")
    
```
so i must add a Router.ts for this page in path `src/app/name/api/route.tx`, right?
```
export async function GET({ params }: { params: { name: string; lang: string } }) {
    const item = await getPublicData('name', params.name)
    const features = await getPublicData("features")
    const faq = await getPublicData("faq_detail")
  const it = await item.json()
  const fe = await features.json()
  const fa = await faq.json()
 
  return Response.json({ it, fe, fa })
}
```
But how i using the return data in Page? and then would it be necessary to add an api/route.ts file to every server-side page? Would that be too cumbersome?

**Reply by icyJoseph** [reply]
- No. Why would you think you need the route file? Your server component page can get the - data from the db as it renders. That's, somewhat, the point of RSC.

**Reply by PunkFleet** [reply]
In fact this build issue has stuck me for days. I browsed through many tutorials and solutions but still no luck.
I don't quite understand that I'm just getting the backend data according to the recommended way in the official documentation, am I missing something? I will post all my code and any help and advice will be appreciated.

.env.local
```
NEXT_PUBLIC_HOSTNAME=http://localhost:3000
NEXT_PUBLIC_API_URL=http://localhost:8000

AUTH_SECRET=62786xxxxxxxxxxxxxxxxxxxxxx64c

```
getClientUrl.ts
```
const base = process.env.NEXT_PUBLIC_API_URL;

const DIR: { [key: string]: string } = {
    // Portfolio
    name: base + '/portfolio/get-name/',
    features: base + '/settings/get-features',
    //Content
    faq_detail: base + '/content/get-faq-detail/',
}

export default function getClientUrl(type: keyof typeof DIR) {
    return DIR[type];
}
```
getPublicData.ts
```
'use server';

import getClientUrl from "@/lib/data/client-url";

export default async function getPublicData(type: string,  params?:any){
    let url = getClientUrl(type)
    if (params){
        url = getClientUrl(type) + params + '/'
    }
    const response = await fetch(url, {
        method: 'GET',
        headers: {
            'Content-Type': 'application/json',
        },
        cache: 'no-store',
    })
    if (!response.ok) {
        throw new Error("Error fetching.");
      }
    return response.json()
}
```

src/app/name/page.tsx
```

type Props = {
    params: { name: string; lang: string };
};

export async function generateMetadata({ params }: Props, parent: ResolvingMetadata): Promise<Metadata> {
    const item = await getPublicData("name", params.name);
    return {
        title: `${params.name}`,
        keywords: item.seo_title,
        description: item.seo_description,
    };
}
export default async function Page({ params }: { params: { name: string; lang: string } }) {
    const lang = params.lang;
    const dictionary = await getDictionary(lang);
    const text = dictionary.products;
    
    const item = await getPublicData('name', params.name)
    const features = await getPublicData("features")
    const faq = await getPublicData("faq_detail")
    

    return (
        <Suspense fallback={<div>Loading...</div>}
                                                    {features.map((i: any) => (
                                            <Grid key={i.id} container xs={12} alignItems="center">
                                                <Grid xs={2} sx={{ display: "flex" }} justifyContent="center">
                                                    <CheckCircleIcon fontSize="small" color="success" />
                                                </Grid>
                                                <Grid xs={10} textAlign="start">
                                                    <Typography variant="body2">{i.name}</Typography>
                                                </Grid>
                                            </Grid>
                                        ))}
                                        <Grid container xs={6} sm={8} md={10} justifyContent="start" spacing={1}>
                                                {item.category.map((i: any, index: number) => (
                                                    <Grid key={index} xs={12} sm="auto">
                                                        <Link component={NextLink} underline="none" color="inherit" href={{ pathname: "/name", query: { category: i } }}>
                                                            <Chip label={i} variant="outlined" color="primary" />
                                                        </Link>
                                                    </Grid>
                                                ))}
                                            </Grid> 
       </Suspense>
    )
}
```

**Reply by icyJoseph** [reply]
Remove the `use server` directive. That's for server actions rather. You may want to install and use server-only, but first get your stuff working.

**Reply by icyJoseph** [reply]
Another issue is that when the page pre-renders at Build time, you have to make sure that the endpoints are available. If those are within the same app, it will fail.

**Reply by PunkFleet** [reply]
So when I build, does the http://localhost:8000 service have to be turned on? Currently I don't have it turned on and have removed 'use server' but still get the error:
```
info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules
   Collecting page data ...
   Generating static pages (0/37) ...
TypeError: fetch failed
    at node:internal/deps/undici/undici:13193:13
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async globalThis.fetch (/app/.next/server/chunks/1638.js:1:36497)
    at async s (/app/.next/server/app/(client)/[lang]/buy-names/[name]/page.js:1:30604)
    at async Z (/app/.next/server/app/(client)/[lang]/help/page.js:1:20812) {
  [cause]: AggregateError [ECONNREFUSED]: 
      at internalConnectMultiple (node:net:1118:18)
      at afterConnectMultiple (node:net:1685:7)
      at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
    code: 'ECONNREFUSED',
    [errors]: [ [Error], [Error] ]
  }
}
   Generating static pages (9/37) 
TypeError: fetch failed
    at node:internal/deps/undici/undici:13193:13
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async globalThis.fetch (/app/.next/server/chunks/1638.js:1:36497)
    at async s (/app/.next/server/app/(client)/[lang]/buy-names/[name]/page.js:1:30604)
    at async Z (/app/.next/server/app/(client)/[lang]/help/page.js:1:20812) {
  [cause]: AggregateError [ECONNREFUSED]: 
      at internalConnectMultiple (node:net:1118:18)
      at afterConnectMultiple (node:net:1685:7)
      at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
    code: 'ECONNREFUSED',
    [errors]: [ [Error], [Error] ]
  }
}
........
Error occurred prerendering page "/en". Read more: https://nextjs.org/docs/messages/prerender-error
.......

> Export encountered errors on following paths:
        /(client)/[lang]/page: /en
Error: building at STEP "RUN npm run build": while running runtime: exit status 1

```

**Reply by icyJoseph** [reply]
So, if I understand correctly, your Portfolio and Content endpoints are the ones doing the actual work. What is it they do? Why can't you just do that instead of using your get data fetch wrapper?

**Reply by PunkFleet** [reply]
Yes, The 8000 is an api service built with django and all fetch requests are sent to this server, currently I don't have it turned on locally.

**Reply by icyJoseph** [reply]
> So when I build, does the http://localhost:8000 service have to be turned on? Currently I don't have it turned on and have removed 'use server' but still get the error:
> ```
> info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules
>    Collecting page data ...
>    Generating static pages (0/37) ...
> TypeError: fetch failed
>     at node:internal/deps/undici/undici:13193:13
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async globalThis.fetch (/app/.next/server/chunks/1638.js:1:36497)
>     at async s (/app/.next/server/app/(client)/[lang]/buy-names/[name]/page.js:1:30604)
>     at async Z (/app/.next/server/app/(client)/[lang]/help/page.js:1:20812) {
>   [cause]: AggregateError [ECONNREFUSED]: 
>       at internalConnectMultiple (node:net:1118:18)
>       at afterConnectMultiple (node:net:1685:7)
>       at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
>     code: 'ECONNREFUSED',
>     [errors]: [ [Error], [Error] ]
>   }
> }
>    Generating static pages (9/37) 
> TypeError: fetch failed
>     at node:internal/deps/undici/undici:13193:13
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>     at async globalThis.fetch (/app/.next/server/chunks/1638.js:1:36497)
>     at async s (/app/.next/server/app/(client)/[lang]/buy-names/[name]/page.js:1:30604)
>     at async Z (/app/.next/server/app/(client)/[lang]/help/page.js:1:20812) {
>   [cause]: AggregateError [ECONNREFUSED]: 
>       at internalConnectMultiple (node:net:1118:18)
>       at afterConnectMultiple (node:net:1685:7)
>       at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
>     code: 'ECONNREFUSED',
>     [errors]: [ [Error], [Error] ]
>   }
> }
> ........
> Error occurred prerendering page "/en". Read more: https://nextjs.org/docs/messages/prerender-error
> .......
> 
> > Export encountered errors on following paths:
>         /(client)/[lang]/page: /en
> Error: building at STEP "RUN npm run build": while running runtime: exit status 1
> 
> ```


Yes, because it is pre-rendering the page, and it literally executes everything the page needs, as declared by you.

Try adding, `export const dynamic = 'force-dynamic'` to defer to, on-demand rendering

**Reply by PunkFleet** [reply]
The 8000api server built with django also includes a request interface for the administrator panel, which is separate from the front-end project that the user can access.

**Reply by icyJoseph** [reply]
> Yes, The 8000 is an api service built with django and all fetch requests are sent to this server, currently I don't have it turned on locally. 

Then you'll need to defer to runtime/on-demand rendering with force-dynamic

**Reply by PunkFleet** [reply]
Thank you sooooooooo much for help , @icyJoseph .😘😘😘
finnally, i build success.
first i opend localhost:8000 server, then i add `export const dynamic = 'force-dynamic'` for every server page and remove 'use server' at `getPublicData.ts`.

But for production, does this mean that I have to get the backend django deployment live before deploying nextjs so that I can deploy it successfully after the .env file changes the address to the real domain? Also in the future would you consider adding a change to const dynamic = 'force-dynamic' in nextjs.config.mjs? Because it seems like adding one of these commands per page is a bit redundant.

**Reply by icyJoseph** [reply]
Yeah you need to make the effort to sync your next.js env values with the production values.

Perhaps l, you'd benefit from this experimental feature, Partial pre-rendering: https://nextjs.org/learn/dashboard-app/partial-prerendering

First, though, make sure everything works in production.

**Reply by PunkFleet** [reply]
> works

Yes, It's a completely different build than react, and I still have a lot of new things to learn. But for now, I feel like nextjs is an amalgamation of react and traditional web applications. It's secure and meets modern build standards.
Thanks again for the generous help!🥰
